elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/QueryEngine.java

```diff
@@ -0,0 +1,161 @@
+ *          +----------------+                    |
+ *                 |                        +-----v-----+
+ *             +---v----+                   |           |
+ *             |        |                   |  NoSQL DB |
+ *             | SQL DB |                   |           |
+ *             |        |                   +-----------+
+ *             +--------+
+ * </pre>
+ * Implementor must assume that {@link DataStoreTransaction} will never keep reference to any internal state of a
+ * {@link QueryEngine} object. This ensures the plugability of various {@link QueryEngine} implementations.
+ * <p>
+ * A concrete and complete {@link QueryEngine} must also have the corresponding implementations of
+ * <ol>
+ *     <li> a {@link Query} that can be understood by a particular persistent storage or storage client
+ *     <li> a {@link WhereFilterExpression} that maps an Elide WHERE filter clause to an equivalent query constraint,
+ *          which can be interpreted by the particular persistent storage
+ *     <li> a {@link HavingFilterExpression} that maps an Elide WHERE filter clause to an equivalent query constraint,
+ *          which can be interpreted by the particular persistent storage
+ *     <li> one or more {@link FilterExpressionVisitor}s that can split an arbitraty {@link FilterExpression} into
+ *          aforementioned {@link WhereFilterExpression} and {@link HavingFilterExpression}
+ * </ol>
+ */
+public interface QueryEngine {
+
+    /**
+     * Executes the specified {@link Query} against the persistent storage.
+     *
+     * @param query  The query customized for a particular persistent storage or storage client
+     *
+     * @return query results
+     */
+    Iterable<Object> executeQuery(Query query);
+
+    /**
+     * Parses request to generate a {@link Query} object to be executed by this {@link QueryEngine}.
+     * <p>
+     * This method calls {@link #buildQuery(Class, Optional, Optional, Optional, Optional, RequestScope)}. If there is
+     * a filter associated with the query, the filter is splitted into a WHERE and a HAVING filter by calling
+     * {@link #extractWhereExpression(Optional)} and {@link #extractHavingExpression(Optional)}, respectively.
+     *
+     * @param entityClass  The type of the entity involved in this query
+     * @param filterExpression  Filters that can be evaluated in the data store. It is optional for the data store to
+     * attempt evaluation.
+     * @param sorting  Sorting which can be pushed down to the data store.
+     * @param pagination  Pagination which can be pushed down to the data store.
+     * @param scope  Contains request level metadata.
+     *
+     * @return a optimized query consumed by the data store
+     */
+    default Query buildQuery(
+            Class<?> entityClass,
+            Optional<FilterExpression> filterExpression,
+            Optional<Sorting> sorting,
+            Optional<Pagination> pagination,
```

Comment:
We are missing Metrics, Dimensions (for groupings), and TimeDimensions (also for grouping).